From a976e0643af5d8295ed5dbd73167d752f3e3a947 Mon Sep 17 00:00:00 2001
From: Martin Fick <mfick@codeaurora.org>
Date: Fri, 20 Aug 2010 16:44:39 -0600
Subject: [PATCH] Allow Access rights to be edited by clicking on them.

When an acces right is clicked on, load the content of the
right into the "add right" form.  This allows a user to edit
the contents and click add.  Depending on the type of the
modification, the right will either be added, or if
appropriate, it will replace the old right.  Also factor out
the "add right" panel into a new class: AccessRightEditor.
Refactor and clarify much of the operation of this class.

Bug: issue 558
Change-Id: Ifbcf3100a2a629a5b8774b03c436ca0b7e5995d4
---
 .../gerrit/client/admin/AccessRightEditor.java     |  431 ++++++++++++++++++++
 .../google/gerrit/client/admin/AdminConstants.java |    1 +
 .../gerrit/client/admin/AdminConstants.properties  |    1 +
 .../gerrit/client/admin/ProjectAccessScreen.java   |  315 ++-------------
 4 files changed, 468 insertions(+), 280 deletions(-)
 create mode 100644 gerrit-gwtui/src/main/java/com/google/gerrit/client/admin/AccessRightEditor.java

diff --git a/gerrit-gwtui/src/main/java/com/google/gerrit/client/admin/AccessRightEditor.java b/gerrit-gwtui/src/main/java/com/google/gerrit/client/admin/AccessRightEditor.java
new file mode 100644
index 0000000..271b902
--- /dev/null
+++ b/gerrit-gwtui/src/main/java/com/google/gerrit/client/admin/AccessRightEditor.java
@@ -0,0 +1,431 @@
+// Copyright (C) 2008 The Android Open Source Project
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.gerrit.client.admin;
+
+import com.google.gerrit.client.Gerrit;
+import com.google.gerrit.client.rpc.GerritCallback;
+import com.google.gerrit.client.ui.AccountGroupSuggestOracle;
+import com.google.gerrit.common.data.ApprovalType;
+import com.google.gerrit.common.data.ProjectDetail;
+import com.google.gerrit.reviewdb.ApprovalCategory;
+import com.google.gerrit.reviewdb.ApprovalCategoryValue;
+import com.google.gerrit.reviewdb.AccountGroup;
+import com.google.gerrit.reviewdb.Project;
+import com.google.gerrit.reviewdb.RefRight;
+import com.google.gwt.event.dom.client.BlurEvent;
+import com.google.gwt.event.dom.client.BlurHandler;
+import com.google.gwt.event.dom.client.ChangeEvent;
+import com.google.gwt.event.dom.client.ChangeHandler;
+import com.google.gwt.event.dom.client.ClickEvent;
+import com.google.gwt.event.dom.client.ClickHandler;
+import com.google.gwt.event.dom.client.FocusEvent;
+import com.google.gwt.event.dom.client.FocusHandler;
+import com.google.gwt.event.dom.client.KeyCodes;
+import com.google.gwt.event.dom.client.KeyPressEvent;
+import com.google.gwt.event.dom.client.KeyPressHandler;
+import com.google.gwt.event.logical.shared.ValueChangeEvent;
+import com.google.gwt.event.logical.shared.ValueChangeHandler;
+import com.google.gwt.event.logical.shared.HasValueChangeHandlers;
+import com.google.gwt.event.shared.HandlerRegistration;
+import com.google.gwt.user.client.ui.Button;
+import com.google.gwt.user.client.ui.Composite;
+import com.google.gwt.user.client.ui.FlowPanel;
+import com.google.gwt.user.client.ui.Grid;
+import com.google.gwt.user.client.ui.ListBox;
+import com.google.gwt.user.client.ui.SuggestBox;
+import com.google.gwtexpui.globalkey.client.NpTextBox;
+
+public class AccessRightEditor extends Composite
+    implements HasValueChangeHandlers<ProjectDetail> {
+  private Project.NameKey projectKey;
+  private ListBox catBox;
+  private NpTextBox nameTxt;
+  private SuggestBox nameSug;
+  private NpTextBox referenceTxt;
+  private ListBox topBox;
+  private ListBox botBox;
+  private Button addBut;
+  private Button clearBut;
+
+  public AccessRightEditor(final Project.NameKey key) {
+    projectKey = key;
+
+    initWidgets();
+    initCategories();
+
+    final Grid grid = new Grid(5, 2);
+    grid.setText(0, 0, Util.C.columnApprovalCategory() + ":");
+    grid.setWidget(0, 1, catBox);
+
+    grid.setText(1, 0, Util.C.columnGroupName() + ":");
+    grid.setWidget(1, 1, nameSug);
+
+    grid.setText(2, 0, Util.C.columnRefName() + ":");
+    grid.setWidget(2, 1, referenceTxt);
+
+    grid.setText(3, 0, Util.C.columnRightRange() + ":");
+    grid.setWidget(3, 1, topBox);
+
+    grid.setText(4, 0, "");
+    grid.setWidget(4, 1, botBox);
+
+    FlowPanel fp = new FlowPanel();
+    fp.setStyleName(Gerrit.RESOURCES.css().addSshKeyPanel());
+
+    fp.add(grid);
+    fp.add(addBut);
+    fp.add(clearBut);
+    initWidget(fp);
+  }
+
+  protected void initWidgets() {
+    catBox = new ListBox();
+    catBox.addChangeHandler(new ChangeHandler() {
+      @Override
+      public void onChange(final ChangeEvent event) {
+        updateCategorySelection();
+      }
+    });
+
+    nameTxt = new NpTextBox();
+    nameSug = new SuggestBox(new AccountGroupSuggestOracle(), nameTxt);
+    nameTxt.setVisibleLength(50);
+    nameTxt.setText(Util.C.defaultAccountGroupName());
+    nameTxt.addStyleName(Gerrit.RESOURCES.css().inputFieldTypeHint());
+    nameTxt.addFocusHandler(new FocusHandler() {
+      @Override
+      public void onFocus(FocusEvent event) {
+        if (Util.C.defaultAccountGroupName().equals(nameTxt.getText())) {
+          nameTxt.setText("");
+          nameTxt.removeStyleName(Gerrit.RESOURCES.css()
+              .inputFieldTypeHint());
+        }
+      }
+    });
+    nameTxt.addBlurHandler(new BlurHandler() {
+      @Override
+      public void onBlur(BlurEvent event) {
+        if ("".equals(nameTxt.getText())) {
+          nameTxt.setText(Util.C.defaultAccountGroupName());
+          nameTxt.addStyleName(Gerrit.RESOURCES.css().inputFieldTypeHint());
+        }
+      }
+    });
+
+    referenceTxt = new NpTextBox();
+    referenceTxt.setVisibleLength(50);
+    referenceTxt.setText("");
+    referenceTxt.addKeyPressHandler(new KeyPressHandler() {
+      @Override
+      public void onKeyPress(KeyPressEvent event) {
+        if (event.getCharCode() == KeyCodes.KEY_ENTER) {
+          doAddNewRight();
+        }
+      }
+    });
+
+    topBox = new ListBox();
+    botBox = new ListBox();
+
+    addBut = new Button(Util.C.buttonAddProjectRight());
+    addBut.addClickHandler(new ClickHandler() {
+      @Override
+      public void onClick(final ClickEvent event) {
+        doAddNewRight();
+      }
+    });
+
+    clearBut = new Button(Util.C.buttonClearProjectRight());
+    clearBut.addClickHandler(new ClickHandler() {
+      @Override
+      public void onClick(final ClickEvent event) {
+        clear();
+      }
+    });
+  }
+
+  protected void initCategories() {
+    for (final ApprovalType at : Gerrit.getConfig().getApprovalTypes()
+        .getApprovalTypes()) {
+      final ApprovalCategory c = at.getCategory();
+      catBox.addItem(c.getName(), c.getId().get());
+    }
+    for (final ApprovalType at : Gerrit.getConfig().getApprovalTypes()
+        .getActionTypes()) {
+      final ApprovalCategory c = at.getCategory();
+      if (Gerrit.getConfig().getWildProject().equals(projectKey)
+          && ApprovalCategory.OWN.equals(c.getId())) {
+        // Giving out control of the WILD_PROJECT to other groups beyond
+        // Administrators is dangerous. Having control over WILD_PROJECT
+        // is about the same as having Administrator access as users are
+        // able to affect grants in all projects on the system.
+        //
+        continue;
+      }
+      catBox.addItem(c.getName(), c.getId().get());
+    }
+
+    if (catBox.getItemCount() > 0) {
+      catBox.setSelectedIndex(0);
+      updateCategorySelection();
+    }
+  }
+
+  public void enableForm(final boolean on) {
+    final boolean canAdd = on && catBox.getItemCount() > 0;
+    addBut.setEnabled(canAdd);
+    clearBut.setEnabled(canAdd);
+    nameTxt.setEnabled(canAdd);
+    referenceTxt.setEnabled(canAdd);
+    catBox.setEnabled(canAdd);
+    topBox.setEnabled(canAdd);
+    botBox.setEnabled(canAdd);
+  }
+
+  public void clear() {
+    setCat(null);
+    setName(Util.C.defaultAccountGroupName());
+    setReference("");
+  }
+
+  public void load(final RefRight right, final AccountGroup group) {
+    final ApprovalType atype =
+       Gerrit.getConfig().getApprovalTypes().getApprovalType(
+          right.getApprovalCategoryId());
+
+    setCat(atype != null ? atype.getCategory().getName()
+                         : right.getApprovalCategoryId().get() );
+
+    setName(group.getName());
+    setReference(right.getRefPatternForDisplay());
+
+    setRange(atype.getCategory().isRange() ? atype.getValue(right.getMinValue())
+             : null, atype.getValue(right.getMaxValue()) );
+  }
+
+  protected void doAddNewRight() {
+    final ApprovalType at = getApprovalType();
+    ApprovalCategoryValue min = getMin(at);
+    ApprovalCategoryValue max = getMax(at);
+
+    if (at == null || min == null || max == null) {
+      return;
+    }
+
+    final String groupName = nameSug.getText();
+    if ("".equals(groupName)
+        || Util.C.defaultAccountGroupName().equals(groupName)) {
+      return;
+    }
+
+    final String refPattern = referenceTxt.getText();
+
+    addBut.setEnabled(false);
+    Util.PROJECT_SVC.addRight(projectKey, at.getCategory().getId(),
+        groupName, refPattern, min.getValue(), max.getValue(),
+        new GerritCallback<ProjectDetail>() {
+          public void onSuccess(final ProjectDetail result) {
+            addBut.setEnabled(true);
+            nameSug.setText("");
+            referenceTxt.setText("");
+            ValueChangeEvent.fire(AccessRightEditor.this, result);
+          }
+
+          @Override
+          public void onFailure(final Throwable caught) {
+            addBut.setEnabled(true);
+            super.onFailure(caught);
+          }
+        });
+  }
+
+  protected void updateCategorySelection() {
+    final ApprovalType at = getApprovalType();
+
+    if (at == null || at.getValues().isEmpty()) {
+      topBox.setEnabled(false);
+      botBox.setEnabled(false);
+      referenceTxt.setEnabled(false);
+      addBut.setEnabled(false);
+      clearBut.setEnabled(false);
+      return;
+    }
+
+    updateRanges(at);
+  }
+
+  protected void updateRanges(final ApprovalType at) {
+    ApprovalCategoryValue min = null, max = null, last = null;
+
+    topBox.clear();
+    botBox.clear();
+
+    for(final ApprovalCategoryValue v : at.getValues()) {
+      final int nval = v.getValue();
+      final String vStr = String.valueOf(nval);
+
+      String nStr = vStr + ": " + v.getName();
+      if (nval > 0) {
+        nStr = "+" + nStr;
+      }
+
+      topBox.addItem(nStr, vStr);
+      botBox.addItem(nStr, vStr);
+
+      if (min == null || nval < 0) {
+        min = v;
+      } else if (max == null && nval > 0) {
+        max = v;
+      }
+      last = v;
+    }
+
+    if (max == null) {
+      max = last;
+    }
+
+    if (ApprovalCategory.READ.equals(at.getCategory().getId())) {
+      // Special case; for READ the most logical range is just
+      // +1 READ, so assume that as the default for both.
+      min = max;
+    }
+
+    if (! at.getCategory().isRange()) {
+      max = null;
+    }
+
+    setRange(min, max);
+  }
+
+  protected void setCat(final String cat) {
+    if (cat == null) {
+      catBox.setSelectedIndex(0);
+    } else {
+      setSelectedText(catBox, cat);
+    }
+    updateCategorySelection();
+  }
+
+  protected void setName(final String name) {
+    if (Util.C.defaultAccountGroupName().equals(name)) {
+      nameTxt.addStyleName(Gerrit.RESOURCES.css().inputFieldTypeHint());
+    } else {
+      nameTxt.removeStyleName(Gerrit.RESOURCES.css().inputFieldTypeHint());
+    }
+    nameTxt.setText(name);
+  }
+
+  protected void setReference(final String ref) {
+    referenceTxt.setText(ref);
+  }
+
+  protected void setRange(final ApprovalCategoryValue min,
+                          final ApprovalCategoryValue max) {
+    if (min == null || max == null) {
+      botBox.setVisible(false);
+      if (max != null) {
+        setSelectedValue(topBox, "" + max.getValue());
+        return;
+      }
+    } else {
+      botBox.setVisible(true);
+      setSelectedValue(botBox, "" + max.getValue());
+    }
+    setSelectedValue(topBox, "" + min.getValue());
+  }
+
+  private ApprovalType getApprovalType() {
+    int idx = catBox.getSelectedIndex();
+    if (idx < 0) {
+      return null;
+    }
+    return Gerrit.getConfig().getApprovalTypes().getApprovalType(
+             new ApprovalCategory.Id(catBox.getValue(idx)));
+  }
+
+  public ApprovalCategoryValue getMin(ApprovalType at) {
+    final ApprovalCategoryValue top = getTop(at);
+    final ApprovalCategoryValue bot = getBot(at);
+    if (bot == null) {
+      for (final ApprovalCategoryValue v : at.getValues()) {
+        if (0 <= v.getValue() && v.getValue() <= top.getValue()) {
+          return v;
+        }
+      }
+      return at.getMin();
+    }
+
+    if (top.getValue() > bot.getValue()) {
+      return bot;
+    }
+    return top;
+  }
+
+  public ApprovalCategoryValue getMax(ApprovalType at) {
+    final ApprovalCategoryValue top = getTop(at);
+    final ApprovalCategoryValue bot = getBot(at);
+    if (bot == null || bot.getValue() < top.getValue()) {
+      return top;
+    }
+    return bot;
+  }
+
+  protected ApprovalCategoryValue getTop(ApprovalType at) {
+    int idx = topBox.getSelectedIndex();
+    if (idx < 0) {
+      return null;
+    }
+    return at.getValue(Short.parseShort(topBox.getValue(idx)));
+  }
+
+  protected ApprovalCategoryValue getBot(ApprovalType at) {
+    int idx = botBox.getSelectedIndex();
+    if (idx < 0 || ! botBox.isVisible()) {
+      return null;
+    }
+    return at.getValue(Short.parseShort(botBox.getValue(idx)));
+  }
+
+  public HandlerRegistration addValueChangeHandler(
+      final ValueChangeHandler<ProjectDetail> handler) {
+    return addHandler(handler, ValueChangeEvent.getType());
+  }
+
+  public static boolean setSelectedText(ListBox box, String text) {
+    if (text == null) {
+      return false;
+    }
+    for (int i =0 ; i < box.getItemCount(); i++) {
+      if (text.equals(box.getItemText(i))) {
+        box.setSelectedIndex(i);
+        return true;
+      }
+    }
+    return false;
+  }
+
+  public static boolean setSelectedValue(ListBox box, String value) {
+    if (value == null) {
+      return false;
+    }
+    for (int i =0 ; i < box.getItemCount(); i++) {
+      if (value.equals(box.getValue(i))) {
+        box.setSelectedIndex(i);
+        return true;
+      }
+    }
+    return false;
+  }
+}
diff --git a/gerrit-gwtui/src/main/java/com/google/gerrit/client/admin/AdminConstants.java b/gerrit-gwtui/src/main/java/com/google/gerrit/client/admin/AdminConstants.java
index 20de3e3..d1e165b 100644
--- a/gerrit-gwtui/src/main/java/com/google/gerrit/client/admin/AdminConstants.java
+++ b/gerrit-gwtui/src/main/java/com/google/gerrit/client/admin/AdminConstants.java
@@ -31,6 +31,7 @@ public interface AdminConstants extends Constants {
   String buttonChangeGroupType();
   String buttonSelectGroup();
   String buttonAddProjectRight();
+  String buttonClearProjectRight();
   String buttonSaveChanges();
   String useContributorAgreements();
   String useSignedOffBy();
diff --git a/gerrit-gwtui/src/main/java/com/google/gerrit/client/admin/AdminConstants.properties b/gerrit-gwtui/src/main/java/com/google/gerrit/client/admin/AdminConstants.properties
index 28917f1..eaca24d 100644
--- a/gerrit-gwtui/src/main/java/com/google/gerrit/client/admin/AdminConstants.properties
+++ b/gerrit-gwtui/src/main/java/com/google/gerrit/client/admin/AdminConstants.properties
@@ -12,6 +12,7 @@ buttonChangeGroupOwner = Change Owner
 buttonChangeGroupType = Change Type
 buttonSelectGroup = Select
 buttonAddProjectRight = Add Access Right
+buttonClearProjectRight = Clear Form
 buttonSaveChanges = Save Changes
 useContributorAgreements = Require a valid contributor agreement to upload
 useSignedOffBy = Require <a href="http://gerrit.googlecode.com/svn/documentation/2.0/user-signedoffby.html#Signed-off-by" target="_blank"><code>Signed-off-by</code></a> in commit message
diff --git a/gerrit-gwtui/src/main/java/com/google/gerrit/client/admin/ProjectAccessScreen.java b/gerrit-gwtui/src/main/java/com/google/gerrit/client/admin/ProjectAccessScreen.java
index 14449f9..e01ee13 100644
--- a/gerrit-gwtui/src/main/java/com/google/gerrit/client/admin/ProjectAccessScreen.java
+++ b/gerrit-gwtui/src/main/java/com/google/gerrit/client/admin/ProjectAccessScreen.java
@@ -18,7 +18,6 @@ import com.google.gerrit.client.Dispatcher;
 import com.google.gerrit.client.Gerrit;
 import com.google.gerrit.client.rpc.GerritCallback;
 import com.google.gerrit.client.rpc.ScreenLoadCallback;
-import com.google.gerrit.client.ui.AccountGroupSuggestOracle;
 import com.google.gerrit.client.ui.FancyFlexTable;
 import com.google.gerrit.client.ui.Hyperlink;
 import com.google.gerrit.client.ui.SmallHeading;
@@ -27,31 +26,19 @@ import com.google.gerrit.common.data.GerritConfig;
 import com.google.gerrit.common.data.InheritedRefRight;
 import com.google.gerrit.common.data.ProjectDetail;
 import com.google.gerrit.reviewdb.AccountGroup;
-import com.google.gerrit.reviewdb.ApprovalCategory;
 import com.google.gerrit.reviewdb.ApprovalCategoryValue;
 import com.google.gerrit.reviewdb.Project;
 import com.google.gerrit.reviewdb.RefRight;
-import com.google.gwt.event.dom.client.BlurEvent;
-import com.google.gwt.event.dom.client.BlurHandler;
-import com.google.gwt.event.dom.client.ChangeEvent;
-import com.google.gwt.event.dom.client.ChangeHandler;
 import com.google.gwt.event.dom.client.ClickEvent;
 import com.google.gwt.event.dom.client.ClickHandler;
-import com.google.gwt.event.dom.client.FocusEvent;
-import com.google.gwt.event.dom.client.FocusHandler;
-import com.google.gwt.event.dom.client.KeyCodes;
-import com.google.gwt.event.dom.client.KeyPressEvent;
-import com.google.gwt.event.dom.client.KeyPressHandler;
+import com.google.gwt.event.logical.shared.ValueChangeEvent;
+import com.google.gwt.event.logical.shared.ValueChangeHandler;
 import com.google.gwt.user.client.ui.Button;
 import com.google.gwt.user.client.ui.CheckBox;
-import com.google.gwt.user.client.ui.FlowPanel;
-import com.google.gwt.user.client.ui.Grid;
-import com.google.gwt.user.client.ui.ListBox;
+import com.google.gwt.user.client.ui.HTMLTable.Cell;
 import com.google.gwt.user.client.ui.Panel;
-import com.google.gwt.user.client.ui.SuggestBox;
 import com.google.gwt.user.client.ui.VerticalPanel;
 import com.google.gwt.user.client.ui.FlexTable.FlexCellFormatter;
-import com.google.gwtexpui.globalkey.client.NpTextBox;
 import com.google.gwtexpui.safehtml.client.SafeHtml;
 import com.google.gwtexpui.safehtml.client.SafeHtmlBuilder;
 
@@ -65,14 +52,7 @@ public class ProjectAccessScreen extends ProjectScreen {
 
   private RightsTable rights;
   private Button delRight;
-  private Button addRight;
-  private ListBox catBox;
-  private ListBox rangeMinBox;
-  private ListBox rangeMaxBox;
-  private NpTextBox nameTxtBox;
-  private SuggestBox nameTxt;
-  private NpTextBox referenceTxt;
-  private FlowPanel addPanel;
+  private AccessRightEditor rightEditor;
 
   public ProjectAccessScreen(final Project.NameKey toShow) {
     super(toShow);
@@ -99,14 +79,7 @@ public class ProjectAccessScreen extends ProjectScreen {
 
   private void enableForm(final boolean on) {
     delRight.setEnabled(on);
-
-    final boolean canAdd = on && catBox.getItemCount() > 0;
-    addRight.setEnabled(canAdd);
-    nameTxtBox.setEnabled(canAdd);
-    referenceTxt.setEnabled(canAdd);
-    catBox.setEnabled(canAdd);
-    rangeMinBox.setEnabled(canAdd);
-    rangeMaxBox.setEnabled(canAdd);
+    rightEditor.enableForm(on);
   }
 
   private void initParent() {
@@ -119,102 +92,6 @@ public class ProjectAccessScreen extends ProjectScreen {
   }
 
   private void initRights() {
-    addPanel = new FlowPanel();
-    addPanel.setStyleName(Gerrit.RESOURCES.css().addSshKeyPanel());
-
-    final Grid addGrid = new Grid(5, 2);
-
-    catBox = new ListBox();
-    rangeMinBox = new ListBox();
-    rangeMaxBox = new ListBox();
-
-    catBox.addChangeHandler(new ChangeHandler() {
-      @Override
-      public void onChange(final ChangeEvent event) {
-        updateCategorySelection();
-      }
-    });
-    for (final ApprovalType at : Gerrit.getConfig().getApprovalTypes()
-        .getApprovalTypes()) {
-      final ApprovalCategory c = at.getCategory();
-      catBox.addItem(c.getName(), c.getId().get());
-    }
-    for (final ApprovalType at : Gerrit.getConfig().getApprovalTypes()
-        .getActionTypes()) {
-      final ApprovalCategory c = at.getCategory();
-      if (Gerrit.getConfig().getWildProject().equals(getProjectKey())
-          && ApprovalCategory.OWN.equals(c.getId())) {
-        // Giving out control of the WILD_PROJECT to other groups beyond
-        // Administrators is dangerous. Having control over WILD_PROJECT
-        // is about the same as having Administrator access as users are
-        // able to affect grants in all projects on the system.
-        //
-        continue;
-      }
-      catBox.addItem(c.getName(), c.getId().get());
-    }
-
-    addGrid.setText(0, 0, Util.C.columnApprovalCategory() + ":");
-    addGrid.setWidget(0, 1, catBox);
-
-    nameTxtBox = new NpTextBox();
-    nameTxt = new SuggestBox(new AccountGroupSuggestOracle(), nameTxtBox);
-    nameTxtBox.setVisibleLength(50);
-    nameTxtBox.setText(Util.C.defaultAccountGroupName());
-    nameTxtBox.addStyleName(Gerrit.RESOURCES.css().inputFieldTypeHint());
-    nameTxtBox.addFocusHandler(new FocusHandler() {
-      @Override
-      public void onFocus(FocusEvent event) {
-        if (Util.C.defaultAccountGroupName().equals(nameTxtBox.getText())) {
-          nameTxtBox.setText("");
-          nameTxtBox.removeStyleName(Gerrit.RESOURCES.css()
-              .inputFieldTypeHint());
-        }
-      }
-    });
-    nameTxtBox.addBlurHandler(new BlurHandler() {
-      @Override
-      public void onBlur(BlurEvent event) {
-        if ("".equals(nameTxtBox.getText())) {
-          nameTxtBox.setText(Util.C.defaultAccountGroupName());
-          nameTxtBox.addStyleName(Gerrit.RESOURCES.css().inputFieldTypeHint());
-        }
-      }
-    });
-    addGrid.setText(1, 0, Util.C.columnGroupName() + ":");
-    addGrid.setWidget(1, 1, nameTxt);
-
-    referenceTxt = new NpTextBox();
-    referenceTxt.setVisibleLength(50);
-    referenceTxt.setText("");
-    referenceTxt.addKeyPressHandler(new KeyPressHandler() {
-      @Override
-      public void onKeyPress(KeyPressEvent event) {
-        if (event.getCharCode() == KeyCodes.KEY_ENTER) {
-          doAddNewRight();
-        }
-      }
-    });
-
-    addGrid.setText(2, 0, Util.C.columnRefName() + ":");
-    addGrid.setWidget(2, 1, referenceTxt);
-
-    addGrid.setText(3, 0, Util.C.columnRightRange() + ":");
-    addGrid.setWidget(3, 1, rangeMinBox);
-
-    addGrid.setText(4, 0, "");
-    addGrid.setWidget(4, 1, rangeMaxBox);
-
-    addRight = new Button(Util.C.buttonAddProjectRight());
-    addRight.addClickHandler(new ClickHandler() {
-      @Override
-      public void onClick(final ClickEvent event) {
-        doAddNewRight();
-      }
-    });
-    addPanel.add(addGrid);
-    addPanel.add(addRight);
-
     rights = new RightsTable();
 
     delRight = new Button(Util.C.buttonDeleteGroupMembers());
@@ -226,15 +103,18 @@ public class ProjectAccessScreen extends ProjectScreen {
       }
     });
 
+    rightEditor = new AccessRightEditor(getProjectKey());
+    rightEditor.addValueChangeHandler(new ValueChangeHandler<ProjectDetail>() {
+        @Override
+        public void onValueChange(ValueChangeEvent<ProjectDetail> event) {
+          display(event.getValue());
+        }
+      });
+
     add(new SmallHeading(Util.C.headingAccessRights()));
     add(rights);
     add(delRight);
-    add(addPanel);
-
-    if (catBox.getItemCount() > 0) {
-      catBox.setSelectedIndex(0);
-      updateCategorySelection();
-    }
+    add(rightEditor);
   }
 
   void display(final ProjectDetail result) {
@@ -253,7 +133,7 @@ public class ProjectAccessScreen extends ProjectScreen {
 
     rights.display(result.groups, result.rights);
 
-    addPanel.setVisible(result.canModifyAccess);
+    rightEditor.setVisible(result.canModifyAccess);
     delRight.setVisible(rights.getCanDelete());
   }
 
@@ -270,144 +150,9 @@ public class ProjectAccessScreen extends ProjectScreen {
     }
   }
 
-  private void doAddNewRight() {
-    int idx = catBox.getSelectedIndex();
-    final ApprovalType at;
-    ApprovalCategoryValue min, max;
-    if (idx < 0) {
-      return;
-    }
-    at =
-        Gerrit.getConfig().getApprovalTypes().getApprovalType(
-            new ApprovalCategory.Id(catBox.getValue(idx)));
-    if (at == null) {
-      return;
-    }
-
-    idx = rangeMinBox.getSelectedIndex();
-    if (idx < 0) {
-      return;
-    }
-    min = at.getValue(Short.parseShort(rangeMinBox.getValue(idx)));
-    if (min == null) {
-      return;
-    }
-
-    if (at.getCategory().isRange()) {
-      idx = rangeMaxBox.getSelectedIndex();
-      if (idx < 0) {
-        return;
-      }
-      max = at.getValue(Short.parseShort(rangeMaxBox.getValue(idx)));
-      if (max == null) {
-        return;
-      }
-    } else {
-      // If its not a range, the maximum box was disabled.  Use the min
-      // value as the max, and select the min from the category values.
-      //
-      max = min;
-      min = at.getMin();
-      for (ApprovalCategoryValue v : at.getValues()) {
-        if (0 <= v.getValue() && v.getValue() <= max.getValue()) {
-          min = v;
-          break;
-        }
-      }
-    }
-
-    final String groupName = nameTxt.getText();
-    if ("".equals(groupName)
-        || Util.C.defaultAccountGroupName().equals(groupName)) {
-      return;
-    }
-
-    final String refPattern = referenceTxt.getText();
-
-    if (min.getValue() > max.getValue()) {
-      // If the user selects it backwards in the web UI, help them out
-      // by reversing the order to what we would expect.
-      //
-      final ApprovalCategoryValue newMin = max;
-      final ApprovalCategoryValue newMax = min;
-      min = newMin;
-      max = newMax;
-    }
-
-    addRight.setEnabled(false);
-    Util.PROJECT_SVC.addRight(getProjectKey(), at.getCategory().getId(),
-        groupName, refPattern, min.getValue(), max.getValue(),
-        new GerritCallback<ProjectDetail>() {
-          public void onSuccess(final ProjectDetail result) {
-            addRight.setEnabled(true);
-            nameTxt.setText("");
-            referenceTxt.setText("");
-            display(result);
-          }
-
-          @Override
-          public void onFailure(final Throwable caught) {
-            addRight.setEnabled(true);
-            super.onFailure(caught);
-          }
-        });
-  }
-
-  private void updateCategorySelection() {
-    final int idx = catBox.getSelectedIndex();
-    final ApprovalType at;
-    if (idx >= 0) {
-      at =
-          Gerrit.getConfig().getApprovalTypes().getApprovalType(
-              new ApprovalCategory.Id(catBox.getValue(idx)));
-    } else {
-      at = null;
-    }
-
-    if (at == null || at.getValues().isEmpty()) {
-      rangeMinBox.setEnabled(false);
-      rangeMaxBox.setEnabled(false);
-      referenceTxt.setEnabled(false);
-      addRight.setEnabled(false);
-      return;
-    }
-
-    int curIndex = 0, minIndex = -1, maxIndex = -1;
-    rangeMinBox.clear();
-    rangeMaxBox.clear();
-    for (final ApprovalCategoryValue v : at.getValues()) {
-      final String vStr = String.valueOf(v.getValue());
-      String nStr = vStr + ": " + v.getName();
-      if (v.getValue() > 0) {
-        nStr = "+" + nStr;
-      }
-
-      rangeMinBox.addItem(nStr, vStr);
-      rangeMaxBox.addItem(nStr, vStr);
-
-      if (v.getValue() < 0) {
-        minIndex = curIndex;
-      }
-      if (maxIndex < 0 && v.getValue() > 0) {
-        maxIndex = curIndex;
-      }
-
-      curIndex++;
-    }
-    if (ApprovalCategory.READ.equals(at.getCategory().getId())) {
-      // Special case; for READ the most logical range is just
-      // +1 READ, so assume that as the default for both.
-      minIndex = maxIndex;
-    }
-    rangeMinBox.setSelectedIndex(minIndex >= 0 ? minIndex : 0);
-    rangeMaxBox.setSelectedIndex(maxIndex >= 0 ? maxIndex : curIndex - 1);
-    rangeMaxBox.setVisible(at.getCategory().isRange());
-
-    addRight.setEnabled(true);
-  }
-
   private class RightsTable extends FancyFlexTable<RefRight> {
     boolean canDelete;
+    Map<AccountGroup.Id, AccountGroup> groups;
 
     RightsTable() {
       table.setWidth("");
@@ -422,6 +167,13 @@ public class ProjectAccessScreen extends ProjectScreen {
       fmt.addStyleName(0, 3, Gerrit.RESOURCES.css().dataHeader());
       fmt.addStyleName(0, 4, Gerrit.RESOURCES.css().dataHeader());
       fmt.addStyleName(0, 5, Gerrit.RESOURCES.css().dataHeader());
+
+      table.addClickHandler(new ClickHandler() {
+        @Override
+        public void onClick(final ClickEvent event) {
+          onOpenRow(table.getCellForEvent(event).getRowIndex());
+        }
+      });
     }
 
     HashSet<RefRight.Key> getRefRightIdsChecked() {
@@ -436,8 +188,9 @@ public class ProjectAccessScreen extends ProjectScreen {
       return refRightIds;
     }
 
-    void display(final Map<AccountGroup.Id, AccountGroup> groups,
+    void display(final Map<AccountGroup.Id, AccountGroup> grps,
         final List<InheritedRefRight> refRights) {
+      groups = grps;
       canDelete = false;
 
       while (1 < table.getRowCount())
@@ -447,12 +200,17 @@ public class ProjectAccessScreen extends ProjectScreen {
         final int row = table.getRowCount();
         table.insertRow(row);
         applyDataRowStyle(row);
-        populate(row, groups, r);
+        populate(row, r);
+      }
+    }
+    protected void onOpenRow(final int row) {
+      if (row > 0) {
+        RefRight right = getRowItem(row);
+        rightEditor.load(right, groups.get(right.getAccountGroupId()));
       }
     }
 
-    void populate(final int row, final Map<AccountGroup.Id, AccountGroup> groups,
-        final InheritedRefRight r) {
+    void populate(final int row, final InheritedRefRight r) {
       final GerritConfig config = Gerrit.getConfig();
       final RefRight right = r.getRight();
       final ApprovalType ar =
@@ -467,11 +225,8 @@ public class ProjectAccessScreen extends ProjectScreen {
         canDelete = true;
       }
 
-      if (ar != null) {
-        table.setText(row, 2, ar.getCategory().getName());
-      } else {
-        table.setText(row, 2, right.getApprovalCategoryId().get());
-      }
+      table.setText(row, 2, ar != null ? ar.getCategory().getName()
+                                       : right.getApprovalCategoryId().get() );
 
       if (group != null) {
         table.setText(row, 3, group.getName());
-- 
1.7.2.2

